<a href="https://colab.research.google.com/github/LCherop/KaggleCompetitions/blob/main/IEEE-CIS%20Fraud%20Detection%20Chamber/fraud_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#IEEE-CIS Fraud Detection


## Import necessary libraries

In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
import sklearn

##Import Datasets

In [ ]:
identity_df = pd.read_csv('/content/train_identity.csv')
transaction_df = pd.read_csv('/content/train_transaction.csv')
print(identity_df.shape)
print(transaction_df.shape)

(144233, 41)
(590540, 394)


##Explore Datasets

In [ ]:
identity_df.head()

TransactionID  id_01     id_02  id_03  id_04  id_05  id_06  id_07  id_08  \
0        2987004    0.0   70787.0    NaN    NaN    NaN    NaN    NaN    NaN   
1        2987008   -5.0   98945.0    NaN    NaN    0.0   -5.0    NaN    NaN   
2        2987010   -5.0  191631.0    0.0    0.0    0.0    0.0    NaN    NaN   
3        2987011   -5.0  221832.0    NaN    NaN    0.0   -6.0    NaN    NaN   
4        2987016    0.0    7460.0    0.0    0.0    1.0    0.0    NaN    NaN   

   id_09  ...                id_31  id_32      id_33           id_34  id_35  \
0    NaN  ...  samsung browser 6.2   32.0  2220x1080  match_status:2      T   
1    NaN  ...   mobile safari 11.0   32.0   1334x750  match_status:1      T   
2    0.0  ...          chrome 62.0    NaN        NaN             NaN      F   
3    NaN  ...          chrome 62.0    NaN        NaN             NaN      F   
4    0.0  ...          chrome 62.0   24.0   1280x800  match_status:2      T   

  id_36 id_37  id_38  DeviceType                     DeviceInfo  
0     F     T      T      mobile  SAMSUNG SM-G892A Build/NRD90M  
1     F     F      T      mobile                     iOS Device  
2     F     T      T     desktop                        Windows  
3     F     T      T     desktop                            NaN  
4     F     T      T     desktop                          MacOS  

[5 rows x 41 columns]

In [ ]:
identity_df.describe()

TransactionID          id_01          id_02         id_03  \
count   1.442330e+05  144233.000000  140872.000000  66324.000000   
mean    3.236329e+06     -10.170502  174716.584708      0.060189   
std     1.788496e+05      14.347949  159651.816856      0.598231   
min     2.987004e+06    -100.000000       1.000000    -13.000000   
25%     3.077142e+06     -10.000000   67992.000000      0.000000   
50%     3.198818e+06      -5.000000  125800.500000      0.000000   
75%     3.392923e+06      -5.000000  228749.000000      0.000000   
max     3.577534e+06       0.000000  999595.000000     10.000000   

              id_04          id_05          id_06        id_07        id_08  \
count  66324.000000  136865.000000  136865.000000  5155.000000  5155.000000   
mean      -0.058938       1.615585      -6.698710    13.285354   -38.600388   
std        0.701015       5.249856      16.491104    11.384207    26.084899   
min      -28.000000     -72.000000    -100.000000   -46.000000  -100.000000   
25%        0.000000       0.000000      -6.000000     5.000000   -48.000000   
50%        0.000000       0.000000       0.000000    14.000000   -34.000000   
75%        0.000000       1.000000       0.000000    22.000000   -23.000000   
max        0.000000      52.000000       0.000000    61.000000     0.000000   

              id_09  ...          id_17         id_18          id_19  \
count  74926.000000  ...  139369.000000  45113.000000  139318.000000   
mean       0.091023  ...     189.451377     14.237337     353.128174   
std        0.983842  ...      30.375360      1.561302     141.095343   
min      -36.000000  ...     100.000000     10.000000     100.000000   
25%        0.000000  ...     166.000000     13.000000     266.000000   
50%        0.000000  ...     166.000000     15.000000     341.000000   
75%        0.000000  ...     225.000000     15.000000     427.000000   
max       25.000000  ...     229.000000     29.000000     671.000000   

               id_20        id_21        id_22        id_24        id_25  \
count  139261.000000  5159.000000  5169.000000  4747.000000  5132.000000   
mean      403.882666   368.269820    16.002708    12.800927   329.608924   
std       152.160327   198.847038     6.897665     2.372447    97.461089   
min       100.000000   100.000000    10.000000    11.000000   100.000000   
25%       256.000000   252.000000    14.000000    11.000000   321.000000   
50%       472.000000   252.000000    14.000000    11.000000   321.000000   
75%       533.000000   486.500000    14.000000    15.000000   371.000000   
max       661.000000   854.000000    44.000000    26.000000   548.000000   

             id_26         id_32  
count  5163.000000  77586.000000  
mean    149.070308     26.508597  
std      32.101995      3.737502  
min     100.000000      0.000000  
25%     119.000000     24.000000  
50%     149.000000     24.000000  
75%     169.000000     32.000000  
max     216.000000     32.000000  

[8 rows x 24 columns]

In [ ]:
transaction_df.head()

TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987000        0          86400            68.5         W  13926   
1        2987001        0          86401            29.0         W   2755   
2        2987002        0          86469            59.0         W   4663   
3        2987003        0          86499            50.0         W  18132   
4        2987004        0          86506            50.0         H   4497   

   card2  card3       card4  card5  ... V330  V331  V332  V333  V334 V335  \
0    NaN  150.0    discover  142.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
1  404.0  150.0  mastercard  102.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
2  490.0  150.0        visa  166.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
3  567.0  150.0  mastercard  117.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
4  514.0  150.0  mastercard  102.0  ...  0.0   0.0   0.0   0.0   0.0  0.0   

  V336  V337  V338  V339  
0  NaN   NaN   NaN   NaN  
1  NaN   NaN   NaN   NaN  
2  NaN   NaN   NaN   NaN  
3  NaN   NaN   NaN   NaN  
4  0.0   0.0   0.0   0.0  

[5 rows x 394 columns]

In [ ]:
transaction_df.describe()

TransactionID        isFraud  TransactionDT  TransactionAmt  \
count   5.905400e+05  590540.000000   5.905400e+05   590540.000000   
mean    3.282270e+06       0.034990   7.372311e+06      135.027176   
std     1.704744e+05       0.183755   4.617224e+06      239.162522   
min     2.987000e+06       0.000000   8.640000e+04        0.251000   
25%     3.134635e+06       0.000000   3.027058e+06       43.321000   
50%     3.282270e+06       0.000000   7.306528e+06       68.769000   
75%     3.429904e+06       0.000000   1.124662e+07      125.000000   
max     3.577539e+06       1.000000   1.581113e+07    31937.391000   

               card1          card2          card3          card5  \
count  590540.000000  581607.000000  588975.000000  586281.000000   
mean     9898.734658     362.555488     153.194925     199.278897   
std      4901.170153     157.793246      11.336444      41.244453   
min      1000.000000     100.000000     100.000000     100.000000   
25%      6019.000000     214.000000     150.000000     166.000000   
50%      9678.000000     361.000000     150.000000     226.000000   
75%     14184.000000     512.000000     150.000000     226.000000   
max     18396.000000     600.000000     231.000000     237.000000   

               addr1          addr2  ...          V330           V331  \
count  524834.000000  524834.000000  ...  82351.000000   82351.000000   
mean      290.733794      86.800630  ...      0.775874     721.741883   
std       101.741072       2.690623  ...      4.727971    6217.223583   
min       100.000000      10.000000  ...      0.000000       0.000000   
25%       204.000000      87.000000  ...      0.000000       0.000000   
50%       299.000000      87.000000  ...      0.000000       0.000000   
75%       330.000000      87.000000  ...      0.000000       0.000000   
max       540.000000     102.000000  ...     55.000000  160000.000000   

                V332           V333          V334         V335          V336  \
count   82351.000000   82351.000000  82351.000000  82351.00000  82351.000000   
mean     1375.783644    1014.622782      9.807015     59.16455     28.530903   
std     11169.275702    7955.735482    243.861391    387.62948    274.576920   
min         0.000000       0.000000      0.000000      0.00000      0.000000   
25%         0.000000       0.000000      0.000000      0.00000      0.000000   
50%         0.000000       0.000000      0.000000      0.00000      0.000000   
75%        25.000000       0.000000      0.000000      0.00000      0.000000   
max    160000.000000  160000.000000  55125.000000  55125.00000  55125.000000   

                V337           V338           V339  
count   82351.000000   82351.000000   82351.000000  
mean       55.352422     151.160542     100.700882  
std       668.486833    1095.034387     814.946722  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000  
75%         0.000000       0.000000       0.000000  
max    104060.000000  104060.000000  104060.000000  

[8 rows x 380 columns]

##Reduce Memory Usage

In [ ]:
def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage().sum() / 1024**2
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose: print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df

In [ ]:
identity_df= reduce_mem_usage(identity_df)
transaction_df= reduce_mem_usage(transaction_df)


Mem. usage decreased to 25.86 Mb (42.7% reduction)
Mem. usage decreased to 542.35 Mb (69.4% reduction)


##Data Cleaning
###Find and remove columns that do not have corresponing transaction ID info

In [ ]:
# Merge transaction_df and identity_df on transactionID (inner join)
merged_df = transaction_df.merge(identity_df, how='inner', on='TransactionID')

# Now merged_df contains entries with corresponding IDs in both DataFrames

# Select desired columns from merged_df for each DataFrame
clean_transaction_df = merged_df[transaction_df.columns]
clean_identity_df = merged_df[identity_df.columns]
print(clean_identity_df.shape)
print(clean_transaction_df.shape)

(144233, 41)
(144233, 394)


###Find and Drop columns with more that 70% Null Values

In [ ]:
#Define function that will be used to calculate column null percentage
def get_null_percentage(df, threshold=70):
    null_counts = df.isnull().sum()
    null_percentages = (null_counts / len(df)) * 100
    return null_percentages[null_percentages > threshold].index.tolist()

In [ ]:
#Calculate for both dataframes
#Calling the list df_sus_cols - because  they are suspicious; why do they have so many NaN
identity_sus_cols = get_null_percentage(clean_identity_df)
transaction_sus_cols = get_null_percentage(clean_transaction_df)
print(identity_sus_cols)
print(transaction_sus_cols)

['id_07', 'id_08', 'id_21', 'id_22', 'id_23', 'id_24', 'id_25', 'id_26', 'id_27']
['dist1', 'dist2', 'D2', 'D3', 'D5', 'D7', 'D11', 'M1', 'M2', 'M3', 'M5', 'M6', 'M7', 'M8', 'M9', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11']


In [ ]:
#Drop the columns with >70% null values
clean_identity_df.drop(columns = identity_sus_cols, inplace=True)
clean_transaction_df.drop(columns = transaction_sus_cols, inplace=True)

<ipython-input-28-c8814851cb0c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_identity_df.drop(columns = identity_sus_cols, inplace=True)
<ipython-input-28-c8814851cb0c>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_transaction_df.drop(columns = transaction_sus_cols, inplace=True)


###Impute missing int values

In [ ]:
num_id_cols = clean_identity_df.select_dtypes(include=['int']).columns
num_trans_cols = clean_transaction_df.select_dtypes(include=['int']).columns


for col in num_id_cols:
  clean_identity_df[col].fillna(clean_identity_df[col].mean(), inplace=True)

for col in num_trans_cols:
  clean_transaction_df[col].fillna(clean_transaction_df[col].mean(), inplace=True)

In [ ]:
df_train = clean_transaction_df.merge(clean_identity_df, how='inner', on='TransactionID')
df_train.head()

TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987004        0          86506          50.000         H   4497   
1        2987008        0          86535          15.000         H   2803   
2        2987010        0          86549          75.875         C  16496   
3        2987011        0          86555          16.500         C   4461   
4        2987016        0          86620          30.000         H   1790   

   card2  card3       card4  card5  ...                id_31  id_32  \
0  514.0  150.0  mastercard  102.0  ...  samsung browser 6.2   32.0   
1  100.0  150.0        visa  226.0  ...   mobile safari 11.0   32.0   
2  352.0  117.0  mastercard  134.0  ...          chrome 62.0    NaN   
3  375.0  185.0  mastercard  224.0  ...          chrome 62.0    NaN   
4  555.0  150.0        visa  226.0  ...          chrome 62.0   24.0   

       id_33           id_34 id_35  id_36  id_37  id_38  DeviceType  \
0  2220x1080  match_status:2     T      F      T      T      mobile   
1   1334x750  match_status:1     T      F      F      T      mobile   
2        NaN             NaN     F      F      T      T     desktop   
3        NaN             NaN     F      F      T      T     desktop   
4   1280x800  match_status:2     T      F      T      T     desktop   

                      DeviceInfo  
0  SAMSUNG SM-G892A Build/NRD90M  
1                     iOS Device  
2                        Windows  
3                            NaN  
4                          MacOS  

[5 rows x 399 columns]

In [ ]:
#Identify and remove target variable
y = df_train.pop('isFraud')

In [ ]:
df_train.shape

(144233, 399)

In [48]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144233 entries, 0 to 144232
Columns: 398 entries, TransactionID to DeviceInfo
dtypes: float16(329), float32(45), int16(1), int32(2), object(21)
memory usage: 139.8+ MB


###Impute Categorical Data

In [37]:
def get_unique_counts_for_categorical_data(df):
  """
  This function analyzes categorical columns in a DataFrame and returns a dictionary
  containing the number of unique values for each column.

  Args:
      df (pandas.DataFrame): The DataFrame containing the data.

  Returns:
      dict: A dictionary where keys are column names and values are the number of unique items.
  """
  categorical_cols = df.select_dtypes(include=['object']).columns
  unique_counts = {}
  for col in categorical_cols:
    unique_counts[col] = df[col].nunique()  # Count unique values using nunique()
  return unique_counts

unique_counts = get_unique_counts_for_categorical_data(df_train)

for col, count in unique_counts.items():
  print(f"Column '{col}': {count} unique items")

Column 'ProductCD': 4 unique items
Column 'card4': 4 unique items
Column 'card6': 3 unique items
Column 'P_emaildomain': 59 unique items
Column 'R_emaildomain': 60 unique items
Column 'M4': 3 unique items
Column 'id_12': 2 unique items
Column 'id_15': 3 unique items
Column 'id_16': 2 unique items
Column 'id_28': 2 unique items
Column 'id_29': 2 unique items
Column 'id_30': 75 unique items
Column 'id_31': 130 unique items
Column 'id_33': 260 unique items
Column 'id_34': 4 unique items
Column 'id_35': 2 unique items
Column 'id_36': 2 unique items
Column 'id_37': 2 unique items
Column 'id_38': 2 unique items
Column 'DeviceType': 2 unique items
Column 'DeviceInfo': 1786 unique items


Categorical Data is varied in number, from 2 to 1786. Feature Engineering may be necessary. But first, I would like to get rid of those null values.

###Impute Categorical Data using Model Imputation

In [50]:
!pip install scikit-learn

In [49]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.impute import SimpleImputer

ImportError: cannot import name '_fit_context' from 'sklearn.base' (/usr/local/lib/python3.10/dist-packages/sklearn/base.py)